### PR TITLE
added sync plugin

### PR DESCRIPTION
**Pull request by acaldas**
https://github.com/ripe-tech/ripe-core/issues/3168

### CODE REVIEW

File: src/js/base/ripe.js [hunk 1]
```diff
@@ -30,15 +30,29 @@ ripe.Ripe.prototype.init = function(brand, model, options) {
     this.useDefaults = this.options.useDefaults === undefined ? !this.noDefaults : this.options.useDefaults;
     this.noCombinations = this.options.noCombinations === undefined ? false : this.options.noCombinations;
     this.useCombinations = this.options.useCombinations === undefined ? !this.noCombinations : this.options.useCombinations;
+    this.useSync = this.options.useSync === undefined ? false : this.options.useSync;
```
Comment (joamag) [inline src/js/base/ripe.js]: the best option is to instantiate the sync plugin and then add the plugin to the ripe instance, this way there's no relationship between sync and the Ripe instance

File: src/python/ripe_demo/static/js/ripe.js [hunk 2]
```diff
@@ -584,6 +607,36 @@ ripe.Ripe.prototype._getMaskURL = function(options) {
...
+ripe.plugins = ripe.plugins || {};
+
+ripe.plugins.Sync = function(owner, rules, options) {
+    options = options || {};
+    this.owner = owner;
+    this.rules = rules;
+
+    // binds to the pre parts event so the parts can be changed
+    // so that they comply with the product's sync rules
+    this.owner.bind("pre_parts", function(parts, newPart) {
```
Comment (joamag) [inline src/python/ripe_demo/static/js/ripe.js]: :+1:

File: src/js/base/ripe.js (hunk 1 above)
Comment (acaldas) [reply src/js/base/ripe.js]: How should the plugins be exported?
Currently only the Ripe constructor is being exported.

```javascript
var Ripe = ripe.Ripe;
```

File: src/js/plugins/sync.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
+ripe.Ripe.plugins.SyncPlugin = function(rules, options) {
+    options = options || {};
+    this.rules = rules;
+}
+
+ripe.Ripe.plugins.SyncPlugin.prototype = Object.create(ripe.Ripe.plugins.Plugin.prototype);
+
+ripe.Ripe.plugins.SyncPlugin.prototype.setOwner = function(owner) {
```
Comment (joamag) [inline src/js/plugins/sync.js]: call this start() instead

Comment (joamag) [reply src/js/plugins/sync.js]: better... call it `register(owner)`

File: src/js/base/ripe.js [hunk 4]
```diff
@@ -51,6 +52,7 @@ ripe.Ripe.prototype.init = function(brand, model, options) {
         this.parts = result;
         this.ready = true;
         this.update();
+        this.trigger("pre_parts", this.parts);
         this.trigger("parts", this.parts);
```
Comment (joamag) [inline src/js/base/ripe.js]: add also post_parts :)

File: src/js/base/ripe.js [hunk 5]
```diff
@@ -48,10 +49,15 @@ ripe.Ripe.prototype.init = function(brand, model, options) {
     };
     loadParts.call(this, function(result) {
         result = result || this.parts;
-        this.parts = result;
+        var parts = [];
+        for (var name in result) {
+            var part = result[name];
+            parts.push([name, part.material, part.color]);
+        }
+        this.setParts(parts);
```
Comment (joamag) [inline src/js/base/ripe.js]: why are we changing to an array from the previous object structure ?

Comment (acaldas) [reply src/js/base/ripe.js]: setParts was not being used to set the parts here and it accepts an array. This is to keep the part events consistent

File: src/js/base/ripe.js [hunk 6]
```diff
@@ -48,10 +49,15 @@ ripe.Ripe.prototype.init = function(brand, model, options) {
...
-        this.parts = result;
+        var parts = [];
+        for (var name in result) {
+            var part = result[name];
+            parts.push([name, part.material, part.color]);
+        }
+        this.setParts(parts);
+
         this.ready = true;
         this.update();
```
Comment (joamag) [inline src/js/base/ripe.js]: not required

File: src/js/base/ripe.js [hunk 7]
```diff
@@ -48,10 +49,15 @@ ripe.Ripe.prototype.init = function(brand, model, options) {
     };
     loadParts.call(this, function(result) {
         result = result || this.parts;
-        this.parts = result;
+        var parts = [];
```
Comment (joamag) [inline src/js/base/ripe.js]: create a method that converts part_m (object) into part_l (array)

File: src/js/ripe.js [hunk 8]
```diff
@@ -271,6 +291,7 @@ ripe.Ripe.prototype.setParts = function(update, noUpdate) {
 
     this.update();
     this.trigger("parts", this.parts);
+    this.trigger("post_parts", this.parts);
```
Comment (joamag) [inline src/js/ripe.js]: remove post_parts

File: src/js/ripe.js [hunk 9]
```diff
@@ -246,17 +252,31 @@ ripe.Ripe.prototype.load = function() {
...
     value.material = material;
     value.color = color;
     this.parts[part] = value;
+
+    var newPart = {
+        name: part,
+        material: material,
+        color: color
+    };
+    this.trigger("pre_parts", this.parts, newPart);
```
Comment (joamag) [inline src/js/ripe.js]: calling this event only "part" and pass only the newPart argument

File: src/python/ripe_demo/static/js/ripe.js [hunk 10]
```diff
@@ -584,6 +605,47 @@ ripe.Ripe.prototype._getMaskURL = function(options) {
...
+            // iterates through the parts of the rule and
+            // sets their material and color to be the same
+            // of the reference part
+            for (var index = 0; index < rule.length; index++) {
+                var _part = rule[index];
+                parts[_part].material = value.material;
+                parts[_part].color = value.color;
+            }
+        }
+    }.bind(this));
```
Comment (joamag) [inline src/python/ripe_demo/static/js/ripe.js]: `this.owner.set_parts(this.parts)`

File: src/python/ripe_demo/static/js/ripe.js [hunk 11]
```diff
@@ -246,12 +248,30 @@ ripe.Ripe.prototype.load = function() {
...
+    this.plugins.push(plugin);
+};
+
+ripe.Ripe.prototype.removePlugin = function(plugin) {
+    plugin.unregister(this);
+    this.plugins.splice(this.plugins.indexOf(plugin), 1);
+};
+
 ripe.Ripe.prototype.setPart = function(part, material, color, noUpdate) {
-    var parts = this.parts || {};
```
Comment (joamag) [inline src/python/ripe_demo/static/js/ripe.js]: separate this into setPArt and _setPart that is re-used by setParts

File: src/js/base/ripe.js [hunk 12]
```diff
@@ -101,6 +98,14 @@ ripe.Ripe.prototype.setParts = function(update, noUpdate) {
     this.trigger("parts", this.parts);
 };
 
+ripe.Ripe.prototype._setPart = function(part, material, color) {
```
Comment (joamag) [inline src/js/base/ripe.js]: move it down in the file (first public methods and then private ones)

**Comment by joamag** [comment]
@acaldas one more thing please check the CI problems https://travis-ci.org/ripe-tech/ripe-sdk/builds/367080426?utm_source=github_status&utm_medium=notification

**Pull request merged**